src/Control/Effect/Accum.hs

```diff
@@ -0,0 +1,63 @@
+{- | An effect allowing writes to an accumulated quantity alongside a computed value,
+and reads from the accumulator.
+An 'Accum' @w@ effect keeps track of a monoidal datum of type @w@ and strictly appends to that monoidal value with the 'add' effect.
+Previous writes to that value can be read with the 'look' effect.
+
+Predefined carriers:
+
+* "Control.Carrier.Accum.Church"
+* "Control.Carrier.Accum.Strict". (A lazy carrier is not provided due to the inherent space leaks associated with lazy accumulation monads, similar to lazy writer monads.)
+* "Control.Monad.Trans.Accum"
+* If 'Accum' @w@ is the last effect in a stack, it can be interpreted to a function @w -> (w, a)@ given some result type @a@ and the presence of a 'Monoid' instance for @w@.
+
+-- | @since 1.0.3.0 -- FIXME Is this right?
+-}
+
+-- FIXME Why not haddockish single quotes?
+
+module Control.Effect.Accum
+( -- * Accumulation effect
+  Accum(..)
+, add
+, look
+, looks
+  -- * Re-exports
+, Algebra
+, Has
+, run
+) where
+
+import Control.Algebra
+-- FIXME Why have an internal module if we reexport everything?
+import Control.Effect.Accum.Internal (Accum(..))
+
+
+-- | Write a value to the log.
+--
+-- @
+-- runAccum ('add' w '>>' m) = 'Data.Bifunctor.first' ('mappend' w) '<$>' runAccum m
+-- @
+--
+-- @since 1.0.3.0 -- FIXME Is this right?
+add :: Has (Accum w) sig m => w -> m ()
+add w = send (Add w)
+{-# INLINE add #-}
+
+-- | Look up the previous accumulation
+--
+-- @
+-- runAccum ('look' m) = 'fmap' ('fst' 'Control.Arrow.&&&' 'id') (runAccum m)
+-- @ -- FIXME This law is wrong of course
```

Comment:
This is an interesting case, as I think getting the laws right will also help us design the tests. (Note that the existing tests are mostly just implementing laws, parameterized by some handler.)

I think we could phrase it mostly simply as something like this:

```haskell
snd <$> runAccum (m >> look) = fst <$> runAccum m
```

This:

- is trivially true for `m = pure x` (for any x), since that doesn’t accumulate anything;
- covers `runAccum look` by the above & the laws for `>>`;
- covers any actions containing `add x` (for any x) by the existing `add` law.

Reply:
I found some simpler laws even:

Recall that `runAccum :: w -> AccumC w m a -> m (w, a)`, i.e. the initial state is always passed to `runAccum` (following the convention of http://hackage.haskell.org/package/transformers-0.5.6.2/docs/Control-Monad-Trans-Accum.html#v:runAccum).

Then we expect the following to hold:

```haskell
runAccum w0 (add w >> m) = runAccum (w0 <> w) m
runAccum w (look >>= continuation) = runAccum w (continuation w)
```
These laws have some simple special cases:
```haskell
runAccum _ (add w) = return (w, ())
runAccum w look = return (empty, w)
```

I implemented all these laws as tests, and they pass.